shell/platform/common/client_wrapper/include/flutter/basic_message_channel.h

```diff
@@ -77,7 +79,7 @@ class BasicMessageChannel {
     BinaryMessageHandler binary_handler = [handler, codec, channel_name](
                                               const uint8_t* binary_message,
                                               const size_t binary_message_size,
-                                              BinaryReply binary_reply) {
+                                              const BinaryReply& binary_reply) {
```

Comment:
Is `BinaryReply` not movable? It looks like it should be: https://github.com/flutter/engine/blob/28f14e6eec4f4dbe6d678f758ca7ff4654473f54/shell/platform/common/client_wrapper/include/flutter/binary_messenger.h#L16-L17

Reply:
A const reference is not copied, it essentially accomplishes the same thing

Reply:
Copying my message from: https://discord.com/channels/608014603317936148/608021010377080866/1153752411614220310

> The reference argument doesn't copy, but doesn't assigning the reference argument to a non-reference instance field result in a copy?

Reply:
There's more potential argument for supporting copy elision here and sprinkling in some `move`s, but setting handlers isn't a common operation, so absent evidence that we need it we should probably just change to the reference. I don't think I did this intentionally.

Reply:
BinaryReply and BinaryMessageHandler are std::functions.
https://github.com/flutter/engine/blob/10c480310926d7e04a77ba8a04d321f9d225a545/shell/platform/common/client_wrapper/include/flutter/binary_messenger.h#L13-L24

Given that this is code that deals with async responses/message channels, are we introducing any lifetime issues by changing the third param to pass by reference rather than copy?

I'm actually sort of surprised that the compiler isn't shouting about a signature mismatch between the updated usage and definitions above given that it doesn't look like they were updated in this patch.

Reply:
Side note: If we were on C++23, which we're not, I suppose we'd have the option of [std::move_only_function](https://en.cppreference.com/w/cpp/utility/functional/move_only_function/move_only_function), which in theory restricts things a little more.

Reply:
@stuartmorgan thoughts?

Reply:
To clarify, here's the scenario I'm thinking of:
1. You're in some block of code; let's say a function somewhere. You have some locals.
2. You construct a `BinaryReply` that captures (by reference) one or more of those locals. BinaryReply is a `std::function` so it can capture all it likes, unlike a function ptr.
3. You construct a `BinaryMessageHandler` using that `BinaryReply` which is, as of this change, being passed by reference, and you put that somewhere to be used later.
4. The scope in which you construct this thing ends and everything goes out of scope.

... time passes ...

6. The message handler invokes the BinaryReply and tries to make use of the captured local which is gone now.

Reply:
From a technical standpoint I don't think this changes anything in your scenario. Arguably the by-value signature better communicates that capturing by reference in the fusion is a bad idea though?

Reply:
Ah yeah that's a good point. Those captures are going to be equally broken in a by-copy scenario. I suspect the only code that'll be broken by this is code that's already quite racy and unlikely to exist in practice. The BinaryReply itself would need to go out of scope between the time the BinaryMessageHandler is invoked with it and the time it's actually used, which seems pretty unlikely in practice.

Reply:
I think we're saying slightly different things. My understanding is that there's no actual lifetime change here; AFAIK the by-value capture of the now-reference argument is still going to make a copy at that point.